homeassistant/components/youtube/config_flow.py

```diff
@@ -0,0 +1,162 @@
+"""Config flow for YouTube integration."""
+from __future__ import annotations
+
+from collections.abc import Mapping
+import logging
+from typing import Any
+
+from google.oauth2.credentials import Credentials
+from googleapiclient.discovery import build
+from googleapiclient.http import HttpRequest
+import voluptuous as vol
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_ACCESS_TOKEN, CONF_TOKEN
+from homeassistant.data_entry_flow import FlowResult
+from homeassistant.helpers import config_entry_oauth2_flow
+from homeassistant.helpers.selector import (
+    SelectOptionDict,
+    SelectSelector,
+    SelectSelectorConfig,
+)
+
+from .const import CONF_CHANNELS, CONF_UPLOAD_PLAYLIST, DEFAULT_ACCESS, DOMAIN
+
+
+class OAuth2FlowHandler(
+    config_entry_oauth2_flow.AbstractOAuth2FlowHandler, domain=DOMAIN
+):
+    """Config flow to handle Google OAuth2 authentication."""
+
+    _data: dict[str, Any] = {}
+
+    DOMAIN = DOMAIN
+
+    reauth_entry: ConfigEntry | None = None
+
+    @property
+    def logger(self) -> logging.Logger:
+        """Return logger."""
+        return logging.getLogger(__name__)
+
+    @property
+    def extra_authorize_data(self) -> dict[str, Any]:
+        """Extra data that needs to be appended to the authorize url."""
+        return {
+            "scope": " ".join(DEFAULT_ACCESS),
+            # Add params to ensure we get back a refresh token
+            "access_type": "offline",
+            "prompt": "consent",
+        }
+
+    async def async_step_reauth(self, entry_data: Mapping[str, Any]) -> FlowResult:
+        """Perform reauth upon an API authentication error."""
+        self.reauth_entry = self.hass.config_entries.async_get_entry(
+            self.context["entry_id"]
+        )
+        return await self.async_step_reauth_confirm()
+
+    async def async_step_reauth_confirm(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Confirm reauth dialog."""
+        if user_input is None:
+            return self.async_show_form(step_id="reauth_confirm")
+        return await self.async_step_user()
```

Comment:
This should be for a future PR.

Reply:
Like, all oauth2 integrations in HA already have reauth, so it's difficult to see them as two different things haha

Reply:
Well in that case, I will leave that up to a member to decide on that since it was accepted on first go for Google Mail.

Reply:
I mean if it is easy and it makes sense I am happy to do it

Reply:
Typically we do as small of PRs as possible, so things like option flows and reauth come as separate follow up PRs to increase review speed of the initial PR.  As an example, you mentioned below you were having trouble with the reauth tests and getting them to work.  (Yes obviously the quality is better when reauth works, but it's a balance).

I would suggest: Add the reauth and options flow and their tests in follow up PRs.

I'm happy to help you get reauth working if you have a specific problems/errors you want to discuss since I contributed to the google mail tests.